src/metabase/upload.clj

```diff
@@ -279,6 +281,30 @@
 (defn- file-mime-type [^File file]
   (.detect tika file))
 
+(defn- detect-charset ^String [file]
+  (try
+    (let [detector (UniversalDetector.)
+          buffer   (byte-array 8192)]
+      (with-open [input-stream (io/input-stream file)]
+        (loop []
+          (let [bytes-read (.read input-stream buffer)]
+            (if (pos? bytes-read)
+              (do
+                (.handleData detector buffer 0 bytes-read)
+                (if (.isDone detector)
+                  (.getDetectedCharset detector)
+                  (recur)))
+              (do
+                (.dataEnd detector)
+                (.getDetectedCharset detector)))))))
```

Comment:
This is the height of nitpickery but I'd maybe have something more like


```
(if (pos? bytes-read)
  (do ...
    (when-not (.isDone detector)
      (recur)))
  (.dataEnd detector))
(.getDetectedCharset detector)
```

(in other words, pulling the `.getDetectedCharset` down to simplify some of the if/do logic)